File: backend/app/schemas/evidence_item.py
# CRANE — CRA Norm Engine
# Copyright (C) 2026 Ali Mohammad Hosseini
# SPDX-License-Identifier: AGPL-3.0-or-later
#
# This file is part of CRANE, free software under the GNU Affero General Public
# License v3.0 or later. See <https://www.gnu.org/licenses/>.

from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import EvidenceType
from app.schemas.common import ORMBaseModel, TimestampedRead


class EvidenceItemCreate(BaseModel):
    product_release_id: UUID | None = None
    risk_assessment_id: UUID | None = None
    requirement_mapping_id: UUID | None = None
    supplier_assessment_id: UUID | None = None
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    evidence_type: EvidenceType
    file_path: str | None = Field(default=None, max_length=500)
    external_url: str | None = Field(default=None, max_length=2048)
    uploaded_by_user_id: UUID


class EvidenceItemUpdate(BaseModel):
    product_release_id: UUID | None = None
    risk_assessment_id: UUID | None = None
    requirement_mapping_id: UUID | None = None
    supplier_assessment_id: UUID | None = None
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    evidence_type: EvidenceType | None = None
    file_path: str | None = Field(default=None, max_length=500)
    external_url: str | None = Field(default=None, max_length=2048)
    uploaded_by_user_id: UUID | None = None


class EvidenceItemRead(TimestampedRead):
    product_release_id: UUID | None
    risk_assessment_id: UUID | None
    requirement_mapping_id: UUID | None
    supplier_assessment_id: UUID | None
    title: str
    description: str | None
    evidence_type: EvidenceType
    file_path: str | None
    external_url: str | None
    uploaded_by_user_id: UUID


class EvidenceItemSummaryRead(ORMBaseModel):
    id: UUID
    title: str
    evidence_type: EvidenceType
    product_release_id: UUID | None
    risk_assessment_id: UUID | None
    requirement_mapping_id: UUID | None
    supplier_assessment_id: UUID | None

File: frontend/src/types/evidence-item.ts
export type EvidenceType =
  | "document"
  | "test_report"
  | "sbom"
  | "screenshot"
  | "link"
  | "declaration"
  | "annex_output"
  | "authority_package";

export interface EvidenceItemRead {
  id: string;
  product_release_id: string | null;
  risk_assessment_id: string | null;
  requirement_mapping_id: string | null;
  title: string;
  description: string | null;
  evidence_type: EvidenceType;
  file_path: string | null;
  external_url: string | null;
  uploaded_by_user_id: string;
  created_at: string;
  updated_at: string;
}

export interface EvidenceItemSummaryRead {
  id: string;
  title: string;
  evidence_type: EvidenceType;
  product_release_id: string | null;
  risk_assessment_id: string | null;
  requirement_mapping_id: string | null;
}

export interface EvidenceItemCreate {
  product_release_id?: string | null;
  risk_assessment_id?: string | null;
  requirement_mapping_id?: string | null;
  title: string;
  description?: string | null;
  evidence_type: EvidenceType;
  file_path?: string | null;
  external_url?: string | null;
  uploaded_by_user_id: string;
}

export interface EvidenceItemUpdate {
  product_release_id?: string | null;
  risk_assessment_id?: string | null;
  requirement_mapping_id?: string | null;
  title?: string;
  description?: string | null;
  evidence_type?: EvidenceType;
  file_path?: string | null;
  external_url?: string | null;
  uploaded_by_user_id?: string | null;
}

export type EvidenceItem = EvidenceItemRead;